src/Enzo/enzo_EnzoMethodPmDeposit.cpp

```diff
@@ -166,84 +163,91 @@ void EnzoMethodPmDeposit::compute ( Block * block) throw()
                (ba == be));
 
 
-      enzo_float dens = 0.0;
-      int ia_m = -1;
-
-      // Determine if particle has a constant mass, or if mass varies
-      // and we need to use the mass attribute
-      if (particle.has_constant (it,"mass")){
-
-        const int ic_mass = particle.constant_index (it, "mass");
-        dens = *((enzo_float *)(particle.constant_value (it,ic_mass)));
-      } else {
-        dens = 1.0;
-        ia_m = particle.attribute_index(it, "mass");
-      }
-
-      // Scale mass by volume if particle value is mass instead of density
-      // Required for Cosmology ("mass" is mass)
-      // Not for Collapse ("mass" is density)
-      if (cosmology) {
-        dens *= std::pow(2.0,rank*level);
-      }
-
-      // Accumulated single velocity array for Baryon deposit
-
+      // Loop over batches
       for (int ib=0; ib<particle.num_batches(it); ib++) {
-
-        int np = particle.num_particles(it,ib);
-
-        // Find the mass of each particle. If cosntant generate array of
-        // constant values in order to simply code below
-        enzo_float * pdens = NULL;
-        if (ia_m > 0){
-          pdens = (enzo_float *) particle.attribute_array( it, ia_m, ib);
-        } else {
-          pdens = new enzo_float[np];
-          std::fill_n(pdens, np, 1.0);
+	// Pointer to array of particle masses
+	enzo_float * pmass = NULL;
+        const int np = particle.num_particles(it,ib);
+
+	// Particle masses can be either a constant value for all particles,
+	// or an attribrute array
+	if (particle.is_constant (it,"mass")){
+
+	  // Check if particle also has a mass attribute
+	  ASSERT1("EnzoMethodPmDeposit::compute",
+		  "Particle type %s has both a constant and an attribute called"
+		  "'mass'. Exiting.", particle.type_name(it).c_str(),
+		  particle.is_attribute(it,"mass"));
+	  
+	  // In this case we fill an array of length np with the constant
+	  // value
+	  pmass = new enzo_float[np];
+	  im = particle.constant_index(it,"mass");
+	  for (int ip = 0; ip<np; ip++)
+	    pmass[ip] = *((enzo_float *)(particle.constant_value (it,im)));
```

Comment:
It's a little unfortunate that we allocate, fill, and free the `pmass` buffer for every batch of each particle type with a constant mass. It would probably be better if we did just did this once for each particle type (with a constant mass) using the max batch size of that particle.

I realize that you just copied the existing style, so feel free to ignore this recommendation.

Reply:
Good suggestion, I'll look into it.